### PR TITLE
MTD reconstruction: propagation of track time of flight uncertainty to vertex reconstruction

### PR DESCRIPTION
**Pull request by noepalm**
#### PR description:

Following up on PR #43918, the newly introduced mass-dependent uncertainty on the time of flight of tracks has been propagated to the following reconstruction steps. This additional contribution to track time uncertainty is now:
 * included in the calculation for the MTD hit-to-track matching $\chi^2_t$ in `TrackExtenderWithMTD`;
 * has been included in the`VertexTimeAlgorithmFromTracksPID` vertex time algorithm for the 3D+t step, which now uses a different track time uncertainty for each mass hypothesis;
 * has been added to the total track time uncertainty in `TOFPIDProducer` (used in the 4D vertex reconstruction step through `DAClusterizerInZT` and `VertexTimeAlgorithmLegacy4D`).

Furthermore, an improved definition of vertex time uncertainty has been introduced in both `VertexTimeAlgorithmFromTracksPID` and `VertexTimeAlgorithmLegacy4D`.
All the mentioned modifications have been discussed and presented in MTD DPG meetings ([1](https://indico.cern.ch/event/1390639/contributions/5846173/attachments/2816111/4916250/MTDDPG_080324_palmeri.pdf), [2](https://indico.cern.ch/event/1395368/contributions/5866090/attachments/2825647/4935848/MTDDPG_220324_palmeri.pdf)).

#### PR validation:

The PR has been validated on TTbar14TeV and TTbar14TeV+PU200 samples, producing results documented in the two aforementioned presentations.

### CODE REVIEW

**Comment by jfernan2** [comment]
please test
(checking if some extra time is being introduced, sorry for the duplicated test)

**Comment by jfernan2** [comment]
please abort 
(test seems stuck)

**Comment by jfernan2** [comment]
please test

**Comment by fabiocos** [comment]
any comment?

**Comment by fabiocos** [comment]
type mtd

File: RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc [hunk 1]
```diff
@@ -306,6 +311,9 @@ namespace {
       tofpid.dterror =
```
Comment (mandrenguyen) [inline RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc:311]: Minor comment, but it seems like you take the sqrt here, only to square this term several times below.  Consider simply using the square to begin with.

File: RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc [hunk 2]
```diff
@@ -321,10 +329,12 @@ namespace {
     if (!addPIDError) {
       //*TODO* deal with heavier nucleons and/or BSM case here?
       float chi2_pi = tofpid.dtchi2;
-      float chi2_k =
-          (tofpid.tmtd - tofpid.dt_k - t_vtx) * (tofpid.tmtd - tofpid.dt_k - t_vtx) / (tofpid.dterror * tofpid.dterror);
-      float chi2_p =
-          (tofpid.tmtd - tofpid.dt_p - t_vtx) * (tofpid.tmtd - tofpid.dt_p - t_vtx) / (tofpid.dterror * tofpid.dterror);
+      float chi2_k = (tofpid.tmtd - tofpid.dt_k - t_vtx) * (tofpid.tmtd - tofpid.dt_k - t_vtx) /
```
Comment (mandrenguyen) [inline RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc:332]: Again very minor, but you could pre-define `tofpid.dterror * tofpid.dterror - tofpid.sigma_dt_pi * tofpid.sigma_dt_pi`
or `tofpid.dterror2 - tofpid.sigma_dt_pi * tofpid.sigma_dt_pi`, if you take into account my comment above. 
Then you could use this in the denominator of both `chi2_k` and `chi2_p`.  Not sure it actually would improve performance, given compiler optimizations, but it arguably makes the code a bit tidier.

Comment (noepalm) [reply RecoMTD/TrackExtender/plugins/TrackExtenderWithMTD.cc]: Thank you for the comments. I have addressed them in the latest commit 6d4465f

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
+1

Failing workflow seems related to L1T

FYI: @cms-sw/l1-l2

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4a2012/38713/runTheMatrixINPUT-results/2500.323_l1DPGNANO130Xrun3/step2_l1DPGNANO130Xrun3.log

**Comment by srimanob** [comment]
+Upgrade

**Comment by rappoccio** [comment]
+1

**Pull request merged**